src/box/alter.cc

```diff
@@ -4417,6 +4426,220 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
+			gc_consumer_unregister(replica->gc);
+		replica->gc = new_gc;
+	} else {
+		if (new_gc != NULL)
+			gc_consumer_unregister(new_gc);
+	}
+	return 0;
+}
+
+/**
+ * The trigger invoked on replace in space _gc_consumers.
+ * Note that due to concurrent nature of transactions it is unsafe
+ * to modify replica->gc right here and in on_rollback, so it is
+ * modified only in on_commit triggers - it is safe because they
+ * are called in order of transactions' serialization.
+ */
+static int
+on_replace_dd_gc_consumers(struct trigger * /* trigger */, void *event)
+{
+	struct txn *txn = (struct txn *)event;
+	RegionGuard region_guard(&fiber()->gc);
+	struct txn_stmt *stmt = txn_current_stmt(txn);
+	struct tuple *old_tuple = stmt->old_tuple;
+	struct tuple *new_tuple = stmt->new_tuple;
+	struct gc_consumer_def *old_def = NULL;
+	struct gc_consumer_def *new_def = NULL;
+	struct tt_uuid *replica_uuid = NULL;
+	if (old_tuple != NULL) {
+		old_def =
+			gc_consumer_def_new_from_tuple(old_tuple, &fiber()->gc);
+		if (old_def == NULL)
+			return -1;
+		replica_uuid = &old_def->uuid;
+	}
+	if (new_tuple != NULL) {
+		new_def =
+			gc_consumer_def_new_from_tuple(new_tuple, &fiber()->gc);
+		if (new_def == NULL)
+			return -1;
+		replica_uuid = &new_def->uuid;
+	}
+	assert(old_def != NULL || new_def != NULL);
+
+	/* Just making sure that both tuples have the same uuid. */
+	assert(old_def == NULL || new_def == NULL ||
+	       tt_uuid_is_equal(&old_def->uuid, &new_def->uuid));
+
+	/*
+	 * We cannot rely on the fact that the replica is still registered
+	 * in-memory because it can be dropped in the same transaction, and
+	 * replica_hash will be updated only on commit, so read row right
+	 * from the _cluster. It's important to lookup by uuid, not id,
+	 * to correctly handle the case when uuid of replica is updated
+	 * and id is the same.
+	 */
+	char key[UUID_STR_LEN + 10];
+	char *key_end = key;
+	key_end = mp_encode_array(key_end, 1);
+	key_end = mp_encode_str0(key_end, tt_uuid_str(replica_uuid));
+	assert((size_t)(key_end - key) < sizeof(key));
+	struct tuple *replica_row;
+	if (box_index_get(BOX_CLUSTER_ID, 1, key, key_end, &replica_row) != 0)
+		return -1;
+	bool replica_is_registered = replica_row != NULL;
+
+	if (!replica_is_registered && new_def != NULL) {
+		diag_set(ClientError, ER_UNSUPPORTED, "gc_consumer",
+			 "create or update without replica");
+		return -1;
+	}
+	if (replica_is_registered && new_def == NULL) {
+		diag_set(ClientError, ER_UNSUPPORTED, "gc_consumer",
+			 "delete while its replica is still registered");
+		return -1;
+	}
+
+	struct gc_consumers_txn_trigger_data *trg_data =
+		xregion_alloc_object(&in_txn()->region,
+				     struct gc_consumers_txn_trigger_data);
+	trg_data->uuid = *replica_uuid;
+	trg_data->new_gc = NULL;
+
+	if (new_def != NULL) { /* INSERT, UPDATE */
+		/* Create new consumer and set it to the replica on commit. */
+		if (new_def->has_vclock) {
+			trg_data->new_gc = gc_consumer_register(
```

Comment:
I found this way the most convenient. These are the cases I thought of:

1. We should invalidate stale consumers. If replica subscribes with a very small vclock (for example, after deleting its xlogs manually), and our `gc.vclock` is ahead, we shouldn't keep any xlogs for it - it will need a rebootstrap soon (or it will fetch xlogs from another replica that hasn't cleaned them up).

2. We should advance consumer forward as late as possible (on commit). If update in `_gc_consumers` is rolled back, but cleanup happened between update and rollback, consumer can become outdated, even if nothing has changed in `_gc_consumers`.

3. We should advance consumer back as soon as possible. Consider the case: we have replica with consumer `{1: 100}`, we advance the consumer back to `{1: 10}`, it becomes outdated (it can happen on subscribe with small vclock), then cleanup is happening - the smallest xlog starts with `{1: 200}` now, and then the update is rolled back and we set the consumer to the previous position (`{1: 100}`), but now it's outdated too. Nothing has changed in our storage, but the consumer just became outdated.

As an alternative, we could advance consumer back on replace without expiration and check if it became outdated only on commit. But it won't help if ENOSPC happened before update and rollback.

Probably, the third case is over-engineering and we could just advance consumer on commit and don't care much about the cases when it's advanced back - anyway, now it can happen only on subscribe, so if a replica is trying to subscribe with smaller vclock, it won't subscribe again with larger one.

Also, we might not invalidate stale consumers (they will pin xlogs until the first ENOSPC). Probably, it's even better because replica can fill gap downloading xlogs from another replica and subscribe to us earlier. But all the corner cases I described remain, but they are much more rare because consumer can become invalidated only on ENOSPC.

So, as for me, keeping two independent consumers (old and new) is the simplest solution - the old consumer is representation of the old tuple, the new one is representation of the new tuple, and while we don't now which tuple will be chosen (commit or rollback), we keep both consumers. Even if one of the consumers becomes outdated, it doesn't affect the other one.

Reply:
Aha, ok, thanks for the elaborate explanation, good job 💪.

Reply:
> Probably, the third case is over-engineering

Yep, it is. In this case the GC could happen just a second before the replica came with this smaller vclock, and it will be lost anyway. Our attempt to prevent it won't change anything, this case is broken by design.

Reply:
> we could just advance consumer on commit

Yes too. I feel like it would simplify a lot.

Reply:
> keeping two independent consumers (old and new) is the simplest solution

Have you tried to code the solution with on-commit update and without consumer recreation?

Reply:
Advance consumers only on commit now.